code/__HELPERS/game.dm

```diff
@@ -213,26 +213,189 @@
+
+	return TRUE
+
+///clean this oranges_ear up for future use
+/mob/oranges_ear/proc/unassign()
+	loc:assigned_oranges_ear = null//trollface. our loc should ALWAYS be a turf, no exceptions. if it isnt then this doubles as an error message ;)
+	loc = null
+	references.Cut()
+
+/**
+ * returns every hearaing movable in view to the turf of source not taking into account lighting
+ * useful when you need to maintain always being able to hear something if a sound is emitted from it and you can see it (and youre in range).
+ * otherwise this is just a more expensive version of get_hearers_in_LOS()
+ *
+ * * view_radius - what radius search circle we are using, worse performance as this increases
+ * * source - object at the center of our search area. everything in get_turf(source) is guaranteed to be part of the search area
+ */
 /proc/get_hearers_in_view(view_radius, atom/source)
 	var/turf/center_turf = get_turf(source)
+	if(!center_turf)
+		return
+
 	. = list()
+
+	if(view_radius <= 0)//special case for if only source cares
+		for(var/atom/movable/target as anything in center_turf)
+			var/list/recursive_contents = target.important_recursive_contents?[RECURSIVE_CONTENTS_HEARING_SENSITIVE]
+			if(recursive_contents)
+				. += recursive_contents
+		return
+
+	var/list/hearables_from_grid = SSspatial_grid.orthogonal_range_search(source, RECURSIVE_CONTENTS_HEARING_SENSITIVE, view_radius)
+
+	if(!length(hearables_from_grid))//we know that something is returned by the grid, but we dont know if we need to actually filter down the output
+		return .
+
+	var/list/assigned_oranges_ears = SSspatial_grid.assign_oranges_ears(hearables_from_grid)
+
+	var/old_luminosity = center_turf.luminosity
+	center_turf.luminosity = 6 //man if only we had an inbuilt dview()
+
+	//this is the ENTIRE reason all this shit is worth it due to how view() works and can be optimized
+	//view() constructs lists of viewed atoms by default and specifying a specific type of atom to look for limits the lists it constructs to those of that
+	//primitive type and then when the view operation is completed the output is then typechecked to only iterate through objects in view with the same
+	//typepath. by assigning one /mob/oranges_ear to every turf with hearable atoms on it and giving them references to each one means that:
+	//1. view() only constructs lists of atoms with the mob primitive type and
+	//2. the mobs returned by view are fast typechecked to only iterate through /mob/oranges_ear mobs, which guarantees at most one per turf
+	//on a whole this can outperform iterating through all movables in view() by ~2x especially when hearables are a tiny percentage of movables in view
+	for(var/mob/oranges_ear/ear in view(view_radius, center_turf))
+		. += ear.references
+		ear.unassign()
+		assigned_oranges_ears -= ear
+
+	for(var/mob/oranges_ear/remaining_ear as anything in assigned_oranges_ears)//we need to clean up our mess
+		remaining_ear.unassign()
```

Comment:
Why not just do all the cleaning on one for loop?

Reply:
thats a good catch it should be faster to do it in the second loop. especially since it doesnt require subtracting from it

Reply:
done